File: src/lib/scoring.ts
import { differenceInDays } from 'date-fns';
import { supabase, Event, User, UserPreferences, RankedEvent, EVENT_CATEGORIES, EventCategory } from './supabase';
import { SCORING_CONFIG } from './config';

export class EventScoringService {
  /**
   * Rank events for a specific user
   * Score = (ClusterMatch × CLUSTER_MATCH_WEIGHT) + (MAX_URGENCY_SCORE - DaysUntilEvent)
   */
  async rankEventsForUser(userId: string, limit: number = 10): Promise<RankedEvent[]> {
    // Fetch user data
    const { data: user, error: userError } = await supabase
      .from('users')
      .select('*')
      .eq('id', userId)
      .single();

    if (userError || !user) {
      throw new Error(`Failed to fetch user: ${userError?.message}`);
    }

    // Fetch user preferences
    const { data: preferences, error: prefsError } = await supabase
      .from('user_preferences')
      .select('*')
      .eq('user_id', userId)
      .single();

    if (prefsError || !preferences) {
      throw new Error(`Failed to fetch preferences: ${prefsError?.message}`);
    }

    // Fetch upcoming events
    const { data: events, error: eventsError } = await supabase
      .from('events')
      .select('*')
      .gte('event_date', new Date().toISOString())
      .order('event_date', { ascending: true })
      .limit(100);

    if (eventsError || !events) {
      throw new Error(`Failed to fetch events: ${eventsError?.message}`);
    }

    // Score each event
    const rankedEvents: RankedEvent[] = events.map((event) => {
      const clusterMatch = this.calculateClusterMatch(event, preferences);
      const urgencyScore = this.calculateUrgencyScore(event);
      const score = clusterMatch * SCORING_CONFIG.CLUSTER_MATCH_WEIGHT + urgencyScore;

      return {
        ...event,
        score,
        cluster_match: clusterMatch,
        urgency_score: urgencyScore,
      };
    });

    // Sort by score and return top N
    rankedEvents.sort((a, b) => b.score - a.score);
    return rankedEvents.slice(0, limit);
  }

  /**
   * Rank events for all users in a specific cluster
   */
  async rankEventsForCluster(clusterId: number, limit: number = 10): Promise<Map<string, RankedEvent[]>> {
    // Fetch all users in the cluster
    const { data: users, error: usersError } = await supabase
      .from('users')
      .select('*')
      .eq('pca_cluster_id', clusterId);

    if (usersError || !users) {
      throw new Error(`Failed to fetch cluster users: ${usersError?.message}`);
    }

    // Rank events for each user
    const results = new Map<string, RankedEvent[]>();
    for (const user of users) {
      try {
        const rankedEvents = await this.rankEventsForUser(user.id, limit);
        results.set(user.id, rankedEvents);
      } catch (error) {
        console.error(`Failed to rank events for user ${user.id}:`, error);
      }
    }

    return results;
  }

  /**
   * Calculate how well an event matches user preferences (0-1 scale)
   */
  private calculateClusterMatch(event: Event, preferences: UserPreferences): number {
    if (!event.category) {
      return 0.5; // Default neutral match
    }

    // Validate that the category exists in EVENT_CATEGORIES
    if (!EVENT_CATEGORIES.includes(event.category as EventCategory)) {
      return 0.5; // Default if category is not valid
    }

    // Safely access the preference value with proper type checking
    const category = event.category as EventCategory;
    const preferenceKey = category as keyof UserPreferences;
    
    // Verify the preference key exists and is a number
    if (preferenceKey in preferences) {
      const preferenceValue = preferences[preferenceKey];
      if (typeof preferenceValue === 'number') {
        return preferenceValue;
      }
    }

    return 0.5; // Default if category not found in preferences
  }

  /**
   * Calculate urgency score based on days until event
   * Returns: MAX_URGENCY_SCORE - daysUntilEvent (capped at 0 minimum)
   */
  private calculateUrgencyScore(event: Event): number {
    const now = new Date();
    const eventDate = new Date(event.event_date);
    const daysUntil = differenceInDays(eventDate, now);

    return Math.max(0, SCORING_CONFIG.MAX_URGENCY_SCORE - daysUntil);
  }

  /**
   * Get event recommendations for a user (alias for rankEventsForUser)
   */
  async getRecommendations(userId: string, limit: number = 10): Promise<RankedEvent[]> {
    return this.rankEventsForUser(userId, limit);
  }
}

File: python/lib/scoring.py
"""
Event Scoring Service for Connect3

Ranks events for users based on:
- Cluster match (how well event matches user preferences)
- Urgency score (days until event)
"""
from datetime import datetime
from typing import List, Dict, Any
from dataclasses import dataclass

from .supabase_client import supabase, RankedEvent, EVENT_CATEGORIES
from .config import SCORING_CONFIG


class EventScoringService:
    """Service for scoring and ranking events for users"""
    
    async def rank_events_for_user(self, user_id: str, limit: int = 10) -> List[RankedEvent]:
        """
        Rank events for a specific user
        Score = (ClusterMatch × CLUSTER_MATCH_WEIGHT) + (MAX_URGENCY_SCORE - DaysUntilEvent)
        """
        # Fetch user data
        result = supabase.table("users").select("*").eq("id", user_id).single().execute()
        if not result.data:
            raise ValueError(f"Failed to fetch user: {user_id}")
        user = result.data
        
        # Fetch user preferences
        result = supabase.table("user_preferences").select("*").eq("user_id", user_id).single().execute()
        if not result.data:
            raise ValueError(f"Failed to fetch preferences for user: {user_id}")
        preferences = result.data
        
        # Fetch upcoming events
        now = datetime.now().isoformat()
        result = supabase.table("events").select("*").gte("event_date", now).order("event_date").limit(100).execute()
        events = result.data or []
        
        # Score each event
        ranked_events = []
        for event in events:
            cluster_match = self._calculate_cluster_match(event, preferences)
            urgency_score = self._calculate_urgency_score(event)
            score = cluster_match * SCORING_CONFIG.CLUSTER_MATCH_WEIGHT + urgency_score
            
            ranked_events.append(RankedEvent(
                id=event["id"],
                title=event["title"],
                event_date=event["event_date"],
                created_at=event.get("created_at", ""),
                updated_at=event.get("updated_at", ""),
                description=event.get("description"),
                location=event.get("location"),
                category=event.get("category"),
                tags=event.get("tags"),
                source_url=event.get("source_url"),
                score=score,
                cluster_match=cluster_match,
                urgency_score=urgency_score
            ))
        
        # Sort by score and return top N
        ranked_events.sort(key=lambda e: e.score, reverse=True)
        return ranked_events[:limit]
    
    async def rank_events_for_cluster(
        self, 
        cluster_id: int, 
        limit: int = 10
    ) -> Dict[str, List[RankedEvent]]:
        """Rank events for all users in a specific cluster"""
        # Fetch all users in the cluster
        result = supabase.table("users").select("*").eq("pca_cluster_id", cluster_id).execute()
        users = result.data or []
        
        # Rank events for each user
        results = {}
        for user in users:
            try:
                ranked_events = await self.rank_events_for_user(user["id"], limit)
                results[user["id"]] = ranked_events
            except Exception as e:
                print(f"Failed to rank events for user {user['id']}: {e}")
        
        return results
    
    def _calculate_cluster_match(self, event: Dict[str, Any], preferences: Dict[str, Any]) -> float:
        """Calculate how well an event matches user preferences (0-1 scale)"""
        category = event.get("category")
        if not category:
            return 0.5  # Default neutral match
        
        # Validate that the category exists in EVENT_CATEGORIES
        if category not in EVENT_CATEGORIES:
            return 0.5  # Default if category is not valid
        
        # Get preference value for this category
        preference_value = preferences.get(category)
        if isinstance(preference_value, (int, float)):
            return float(preference_value)
        
        return 0.5  # Default if category not found in preferences
    
    def _calculate_urgency_score(self, event: Dict[str, Any]) -> float:
        """
        Calculate urgency score based on days until event
        Returns: MAX_URGENCY_SCORE - daysUntilEvent (capped at 0 minimum)
        """
        now = datetime.now()
        event_date = datetime.fromisoformat(event["event_date"].replace("Z", "+00:00"))
        days_until = (event_date - now).days
        
        return max(0, SCORING_CONFIG.MAX_URGENCY_SCORE - days_until)
    
    async def get_recommendations(self, user_id: str, limit: int = 10) -> List[RankedEvent]:
        """Get event recommendations for a user (alias for rank_events_for_user)"""
        return await self.rank_events_for_user(user_id, limit)


# Singleton instance
event_scoring_service = EventScoringService()
